ratelimits/limiter.go

```diff
@@ -0,0 +1,234 @@
+package ratelimits
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jmhodges/clock"
+)
+
+// ErrInvalidCost indicates that the cost specified was <= 0.
+var ErrInvalidCost = fmt.Errorf("invalid cost, must be > 0")
+
+// ErrInvalidCostForCheck indicates that the check cost specified was < 0.
+var ErrInvalidCostForCheck = fmt.Errorf("invalid check cost, must be >= 0")
+
+// ErrInvalidCostOverLimit indicates that the cost specified was > limit.Burst.
+var ErrInvalidCostOverLimit = fmt.Errorf("invalid cost, must be <= limit.Burst")
+
+// ErrBucketAlreadyFull indicates that the bucket already has reached its
+// maximum capacity.
+var ErrBucketAlreadyFull = fmt.Errorf("bucket already full")
+
+// Limiter provides a high-level interface for rate limiting requests by
+// utilizing a leaky bucket-style approach.
+type Limiter struct {
+	// defaults stores default limits by 'name'.
+	defaults limits
+
+	// overrides stores override limits by 'name:id'.
+	overrides limits
+
+	// source is used to store buckets. It must be safe for concurrent use.
+	source source
+	clk    clock.Clock
+}
+
+// NewLimiter returns a new *Limiter. The provided source must be safe for
+// concurrent use. The defaults and overrides paths are expected to be paths to
+// YAML files that contain the default and override limits, respectively. The
+// overrides file is optional, all other arguments are required.
+func NewLimiter(clk clock.Clock, source source, defaults, overrides string) (*Limiter, error) {
+	limiter := &Limiter{source: source, clk: clk}
+
+	var err error
+	limiter.defaults, err = loadAndParseDefaultLimits(defaults)
+	if err != nil {
+		return nil, err
+	}
+
+	if overrides == "" {
+		// No overrides specified, initialize an empty map.
+		limiter.overrides = make(limits)
+		return limiter, nil
+	}
+
+	limiter.overrides, err = loadAndParseOverrideLimits(overrides)
+	if err != nil {
+		return nil, err
+	}
+
+	return limiter, nil
+}
+
+type Decision struct {
+	// Allowed is true if the bucket possessed enough capacity to allow the
+	// request given the cost.
+	Allowed bool
+
+	// Remaining is the number of requests the client is allowed to make before
+	// they're rate limited.
+	Remaining int
```

Comment:
`Burst` is an int64, therefore `Remaining` needs to be an int64 as well, since there can be up to `Burst` tokens remaining in a bucket.